fluss-server/src/test/java/com/alibaba/fluss/server/log/WriterStateManagerTest.java

```diff
@@ -179,36 +182,94 @@ void testFetchSnapshotEmptySnapshot() {
                 .hasMessageContaining(
                         "Out of order batch sequence for writer 1 at offset 2 in "
                                 + "table-bucket TableBucket{tableId=1001, bucket=0}"
                                 + " : 2 (incoming batch seq.), -1 (current batch seq.)");
 
-        append(recoveredMapping, 2L, 0, 2L, 70002);
+        append(recoveredMapping, 2L, 0, 2L, false, 70002);
 
         assertThat(recoveredMapping.activeWriters().size()).isEqualTo(1);
         assertThat(recoveredMapping.activeWriters().values().iterator().next().lastBatchSequence())
                 .isEqualTo(0);
         assertThat(recoveredMapping.mapEndOffset()).isEqualTo(3L);
     }
 
+    @Test
+    void testLoadFromSnapshotRetainExpiredWriters() throws Exception {
+        ManualClock clock = new ManualClock(1000L);
+
+        // 2 seconds to expire the writer.
+        conf.set(ConfigOptions.WRITER_ID_EXPIRATION_TIME, Duration.ofSeconds(2));
+        WriterStateManager stateManager1 =
+                new WriterStateManager(
+                        tableBucket,
+                        logDir,
+                        (int) conf.get(ConfigOptions.WRITER_ID_EXPIRATION_TIME).toMillis());
+
+        long writerId1 = 1L;
+        long writerId2 = 2L;
+
+        append(stateManager1, writerId1, 0, 0L, false, clock.milliseconds());
+        append(stateManager1, writerId2, 0, 1L, false, clock.milliseconds());
+        stateManager1.takeSnapshot();
+        assertThat(stateManager1.activeWriters().size()).isEqualTo(2);
+
+        // trigger clock move to 5000L which means the batches expired.
+        clock.advanceTime(5000L, TimeUnit.MILLISECONDS);
+        // new one new WriterStateManager to mock tabletServer restart and reload from snapshot.
+        WriterStateManager stateManager2 =
+                new WriterStateManager(
+                        tableBucket,
+                        logDir,
+                        (int) conf.get(ConfigOptions.WRITER_ID_EXPIRATION_TIME).toMillis());
+        stateManager2.truncateAndReload(0L, 2L);
+        // all writers are retained.
+        assertThat(stateManager2.activeWriters().size()).isEqualTo(2);
+    }
+
+    @Test
+    void testAppendAnExpiredBatchWithEmptyWriterStatus() throws Exception {
+        ManualClock clock = new ManualClock(5000L);
+
+        // 2 seconds to expire the writer.
+        conf.set(ConfigOptions.WRITER_ID_EXPIRATION_TIME, Duration.ofSeconds(2));
+        WriterStateManager stateManager1 =
+                new WriterStateManager(
+                        tableBucket,
+                        logDir,
+                        (int) conf.get(ConfigOptions.WRITER_ID_EXPIRATION_TIME).toMillis());
+
+        // If we try to append an expired batch with none zero batch sequence, the
+        // OutOfOrderSequenceException will not been throw.
+        append(stateManager1, 1L, 10, 10L, true, clock.milliseconds());
```

Comment:
It will be better to add a check after this that the activeWriters in stateManager is 1, and the last sequence is 10 (use the new batch sequence).